azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,1021 @@
+            if self.target:
+                # remove/overwrite attributes that cannot be changed
+                current_hub_resource = self.discovery.find_resource(self.hub_name, self.rg)
+                if not self.rg:
+                    self.rg = current_hub_resource.additional_properties["resourcegroup"]
+                # location
+                hub_resource["location"] = current_hub_resource.location
+                # sku
+                hub_resource["sku"] = current_hub_resource.sku.serialize()
+                # event hub partitions
+                partition_count = current_hub_resource.properties.event_hub_endpoints["events"].partition_count
+                hub_resource["properties"]["eventHubEndpoints"]["events"]["partitionCount"] = partition_count
+                # enable data residency
+                hub_resource["properties"]["enableDataResidency"] = current_hub_resource.properties.enable_data_residency
+                # features - hub takes care of this but we will do this just incase
+                hub_resource["properties"]["features"] = current_hub_resource.properties.features
+                # TODO check for other props and add them as they pop up
+
+            else:
+                # If there is a system assigned identity endpoint, the ARM deployment may stall rather than failing
+                # outright. So warn and fail before deployment.
+                identity_endpoints = []
+                for endpoint_list in hub_resource["properties"]["routing"]["endpoints"].values():
+                    for endpoint in endpoint_list:
+                        if (
+                            endpoint["authenticationType"] == AuthenticationType.IdentityBased.value
+                            and not endpoint.get("identity")
+                        ):
+                            identity_endpoints.append(endpoint["name"])
+                if len(identity_endpoints) > 0:
+                    raise BadRequestError(
+                        usr_msgs.FAILED_ARM_IDENTITY_ENDPOINT_MSG.format(self.hub_name, identity_endpoints)
+                    )
+
+            hub_resources.append(hub_resource)
+
+            hub_certs = [res for res in hub_state["arm"]["resources"][1:] if res["type"].endswith("certificates")]
+            if len(hub_certs) < len(hub_state["arm"]["resources"]) - 1:
+                logger.warning(usr_msgs.PRIVATE_ENDPOINT_WARNING_MSG)
+
+            for res in hub_certs:
+                res["name"] = self.hub_name + "/" + res["name"].split("/")[1]
+                depends_on = res["dependsOn"][0].split("'")
+                depends_on[3] = self.hub_name
+                res["dependsOn"][0] = "'".join(depends_on)
+            hub_resources.extend(hub_certs)
+            hub_state["arm"]["resources"] = hub_resources
+
+            state_file = f"arm_deployment-{self.hub_name}.json"
+            with open(state_file, "w", encoding='utf-8') as f:
+                json.dump(hub_state["arm"], f)
```

Comment:
Will this still succeed (overwrite the entire file) if this filename happens to exist before this command is run?

Reply:
yes - that is how the w flag works